### DISCUSSION TITLE
Ability to rename os_window?

### DISCUSSION THREAD

**Discussion by mollerhoj** [kovidgoyal/kitty#6317] [General]
Hi, I'd love to be able to rename an existing os_window. I'm on mac osx. Is that possible? Could it be made possible?

**Comment by kovidgoyal** [comment]
OS windows dont have names, tey have ids. What are you referring to? The
title an OS window displays is the title of its current active kitty window.

**Comment by mollerhoj** [comment]
I can set the name of a new os_window with `kitty --title "Some name"`.

It will be displayed on hover in the task bar:
<img width="121" alt="Screenshot 2023-05-31 at 12 22 20" src="https://github.com/kovidgoyal/kitty/assets/649188/12ed3fd4-6905-4cbc-b662-d449686aa7ec">

It will show this name regardless of what the name of the kitty tabs are.

But I don't know how to change the name on an existing window, so I have to create a new one to change the name :-)

**Comment by kovidgoyal** [comment]
On Wed, May 31, 2023 at 03:23:17AM -0700, Jens Dahl Møllerhøj wrote:
> I can set the name of a new os_window with `kitty --title "Some name"`.
> 
> It will be displayed on hover in the task bar:
> <img width="121" alt="Screenshot 2023-05-31 at 12 22 20" src="https://github.com/kovidgoyal/kitty/assets/649188/12ed3fd4-6905-4cbc-b662-d449686aa7ec">
> 
> But I don't know how to change the name on an existing window, so I have to create a new one to change the name :-)

If you want to dynamically change titles dont use --title, instead use
the kitty remote control feature:
https://sw.kovidgoyal.net/kitty/remote-control/

**Reply by mollerhoj** [reply]
As far as I can tell, there is no remote control command to set the os window title. `set-window-title` sets the window title, but it is overwritten by tab titles (That is, it does not set the OS window title / id)

**Comment by kovidgoyal** [comment]
On Wed, May 31, 2023 at 06:12:10AM -0700, Jens Dahl Møllerhøj wrote:
> As far as I can tell, there is no remote control command to set the os window title. `set-window-title` sets the window title, but it is overwritten by tab titles (That is, it does not set the OS window title / id)

There is set-tab0title as well. If you use tabs then use that.

**Reply by mollerhoj** [reply]
Okay, that seems quite unrelated though. Is it correctly understood that the os_window title is unchangeable once set? Is that a limitation from the OS or could kitty be extended to support it?

**Reply by mollerhoj** [reply]
Notice, there is a difference between os-window titles, window titles and tab title. The later two can be set with remote control, but it seems os-window is immutable.

**Comment by kovidgoyal** [comment]
The OS window title is whatever the title of the active tab is. It's not
immutable, unless you use --title. I dont see a compelling use case for
giving OS windows separate titles beyond the existing --title semantics.

**Reply by mollerhoj** [reply]
Since the name of the current OS window is copied over to any new OS window one creates, you quickly end up with a lot of windows with the same name (after having used --title).

Here's kitty's "Window" menu:

![Screenshot 2023-07-11 at 17 36 03](https://github.com/kovidgoyal/kitty/assets/649188/fbd85818-5ee6-4273-bb40-fdb9652bc969)

Would be super nice to be able to rename an OS window some time after creation.

**Reply by kovidgoyal** [reply]
Use set-tab-title and dont use --title. IIRC the OS window will pickup the title of the active tab. If you dont use tabs there is only a single active tab ever, so its title should always be used.

**Reply by mollerhoj** [reply]
I do use tabs extensively unfortunately.

**Reply by mollerhoj** [reply]
Would editable titles be possible to implement? (I could start looking into it) or does the OS not allow renaming? Hm, I guess since they are autoupdated with with the tab name if --title is not used, it should be possible..

**Comment by kovidgoyal** [comment]
On Wed, Jul 12, 2023 at 12:09:20AM -0700, Jens Dahl Møllerhøj wrote:
> Would editable titles be possible to implement? (I could start looking into it) or does the OS not allow renaming?

There is no technical reason they cant be implemented.

**Reply by mollerhoj** [reply]
Would you oppose a PR adding a `kitty @ set-os-window-title` subcommand?

**Comment by kovidgoyal** [comment]
On Wed, Jul 12, 2023 at 04:27:51AM -0700, Jens Dahl Møllerhøj wrote:
> Would you oppose a PR adding a `kitty @ set-os-window-title` command?

Not in principle, but it will depend on the details. IIRC title handling
for OS Windows is fairly involved.

**Reply by mollerhoj** [reply]
Gotcha - I imagine the current logic updating the title with the name of the tab could be reused?

**Comment by kovidgoyal** [comment]
Dont recall would have to check the code.

**Comment by hacker-DOM** [comment]
I've been dreaming about this for a long time https://github.com/kovidgoyal/kitty/issues/6313

Seems like the discussion there got confused between kitty windows and OS windows at the end (#takeaway it's best to specify it explicitly each time).

**Reply by hacker-DOM** [reply]
To be fair, out of popular apps I use, only Chrome allows naming OS windows. Still would ❤️ to see it in kiTTY

**Comment by hacker-DOM** [comment]
temporary hack https://github.com/kovidgoyal/kitty/issues/6313#issuecomment-2392698534
